src/vivarium_gates_nutrition_optimization/components/pregnancy.py

```diff
@@ -16,9 +18,88 @@ def __init__(self, cause, *args, **kwargs):
+            self.event_count_column,
+            "pregnancy_term_outcome",
+            "pregnancy_duration",
+            # "pregnancy_birth_outcome",
+        ]
+
     def setup(self, builder: Builder):
         super().setup(builder)
         self.time_step = builder.time.step_size()
+        self.randomness = builder.randomness.get_stream(self.name)
+        self.partial_term_probs = builder.value.register_value_producer(
+            "partial_term_probabilities",
+            source=builder.lookup.build_table(
+                get_partial_term_probability_data(builder),
+                key_columns=["sex"],
+                parameter_columns=["age", "year"],
+            ),
+        )
+
+    def on_initialize_simulants(self, pop_data: SimulantData) -> None:
+        for transition in self.transition_set:
+            if transition.start_active:
+                transition.set_active(pop_data.index)
+
+        pop_events = self.get_initial_event_times(pop_data)
+        pregnancy_term_outcomes_and_durations = self.sample_pregnancy_terms_and_durations(
+            pop_data
+        )
+        pop_update = pd.concat([pop_events, pregnancy_term_outcomes_and_durations], axis=1)
+        self.population_view.update(pop_update)
+
+    def sample_pregnancy_terms_and_durations(self, pop_data: SimulantData) -> pd.DataFrame:
+        term_outcome_probabilities = self.partial_term_probs(pop_data.index)
+        term_outcome_probabilities = pd.DataFrame(
+            {
+                models.PARTIAL_TERM_OUTCOME: term_outcome_probabilities,
+                models.FULL_TERM_OUTCOME: 1 - term_outcome_probabilities,
+            }
+        )
+        pregnancy_term_outcomes = pd.DataFrame(
+            {
+                "pregnancy_term_outcome": self.randomness.choice(
+                    pop_data.index,
+                    choices=term_outcome_probabilities.columns.to_list(),
+                    p=term_outcome_probabilities,
+                    additional_key="pregnancy_term_outcome",
+                )
+            }
+        )
+
+        term_duration_map = {
+            models.FULL_TERM_OUTCOME: lambda *_: pd.to_timedelta(
+                DURATIONS.FULL_TERM, unit="days"
+            ),
+            models.PARTIAL_TERM_OUTCOME: self.sample_partial_term_durations,
+        }
+
+        for term_length, sampling_func in term_duration_map.items():
+            term_pop = pregnancy_term_outcomes[
+                pregnancy_term_outcomes["pregnancy_term_outcome"] == term_length
+            ].index
+            pregnancy_term_outcomes.loc[term_pop, "pregnancy_duration"] = sampling_func(
+                term_pop
+            )
+
+        return pregnancy_term_outcomes
+
+    def sample_partial_term_durations(self, partial_term_pop: pd.DataFrame) -> pd.Series:
+        low, high = DURATIONS.DETECTION, DURATIONS.PARTIAL_TERM
+        draw = self.randomness.get_draw(partial_term_pop, additional_key="pregnancy_duration")
```

Comment:
So you don't have to risk a potential key collision when we add sampling for full-term duration, it'd probably be better to call the additional key something like `"partial_term_pregnancy_duration"`